mithril-common/src/test_utils/certificate_chain_builder.rs

```diff
@@ -0,0 +1,752 @@
+        &self,
+        context: &CertificateChainBuilderContext,
+        genesis_signer: &ProtocolGenesisSigner,
+    ) -> Certificate {
+        let epoch = context.epoch;
+        let certificate = self.build_base_certificate(context);
+        let next_avk = Self::compute_avk_for_signers(&context.next_fixture.signers_fixture());
+        let next_protocol_parameters = &context.next_fixture.protocol_parameters();
+        let genesis_producer =
+            CertificateGenesisProducer::new(Some(Arc::new(genesis_signer.to_owned())));
+        let genesis_protocol_message = CertificateGenesisProducer::create_genesis_protocol_message(
+            next_protocol_parameters,
+            &next_avk,
+            &epoch,
+        )
+        .unwrap();
+        let genesis_signature = genesis_producer
+            .sign_genesis_protocol_message(genesis_protocol_message)
+            .unwrap();
+
+        CertificateGenesisProducer::create_genesis_certificate(
+            certificate.metadata.protocol_parameters,
+            certificate.metadata.network,
+            certificate.epoch,
+            #[allow(deprecated)]
+            certificate.metadata.immutable_file_number,
+            next_avk,
+            genesis_signature,
+        )
+        .unwrap()
+    }
+
+    fn build_standard_certificate(&self, context: &CertificateChainBuilderContext) -> Certificate {
+        let fixture = context.fixture;
+        let mut certificate = self.build_base_certificate(context);
+        certificate.metadata.signers = fixture.stake_distribution_parties();
+        let mut protocol_message = certificate.protocol_message.clone();
+        protocol_message.set_message_part(
+            ProtocolMessagePartKey::SnapshotDigest,
+            format!("digest-{}", context.index_certificate),
+        );
+        certificate.protocol_message = protocol_message;
+        certificate.signed_message = certificate.protocol_message.compute_hash();
+        let single_signatures = fixture
+            .signers_fixture()
+            .iter()
+            .filter_map(|s| {
+                s.protocol_signer
+                    .sign(certificate.signed_message.as_bytes())
+            })
+            .collect::<Vec<_>>();
+        let clerk = CertificateChainBuilder::compute_clerk_for_signers(&fixture.signers_fixture());
+        let multi_signature = clerk
+            .aggregate(&single_signatures, certificate.signed_message.as_bytes())
+            .unwrap();
+        certificate.signature = CertificateSignature::MultiSignature(
+            certificate.signed_entity_type(),
+            multi_signature.into(),
+        );
+
+        certificate
+    }
+
+    // Returns the chained certificates in reverse order
+    // The latest certificate of the chain is the first in the vector
+    fn compute_chained_certificates(&self, certificates: Vec<Certificate>) -> Vec<Certificate> {
+        let mut certificates_chained: Vec<Certificate> = Vec::new();
```

Comment:
Handle the first certificate add complexity to understand what is done.
It may be easier to understand if we create a function to update hash of the ̀Certificate`.
In the method, we compute twice the hash a certificate. We probably can reuse the hash of the previous_certificate instead of recompute it.

I propose you something like (I would like to avoid cloning certificate but I'm not sure it's possible):
```
   fn update_previous_hash(certificate: &Certificate, previous_hash: String) -> Certificate {
        let mut certificate_new = certificate.clone();
        certificate_new.previous_hash = previous_hash.to_string();
        certificate_new.hash = certificate_new.compute_hash();
        certificate_new
    }

    let mut certificates_chained: Vec<Certificate> = Vec::new();
    certificates
        .iter()
        .fold("".to_string(), |previous_hash, certificate| {
            let certificate_new = update_previous_hash(certificate, previous_hash);
            let new_hash = certificate_new.hash.clone();
            certificates_chained.push(certificate_new);

            new_hash
        });
        ``

Reply:
Absolutely, thanks for the proposal! 
I updated the implementation of the certificate chaining in https://github.com/input-output-hk/mithril/pull/1999/commits/c223548733d72af7094486246d299e7f63e3d5ee :+1: